x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+                // RSA/EC JWKSet present without any RSA/EC signature algorithms
+                throw new SettingsException(
+                    "Settings mismatch. RSA/EC JWKSet is set in ["
+                        + jwkSetConfigKeyPkc
+                        + "], but no RSA/EC signature algorithms are set in ["
+                        + allowedSignatureAlgorithmsConfigKey
+                        + "]"
+                );
+            } else if ((allowedSignatureAlgorithmsPkc.isEmpty() == false) && (isConfiguredJwkSetPkc == false)) {
+                // RSA/EC signature algorithms present without an RSA/EC JWKSet
+                throw new SettingsException(
+                    "Settings mismatch. RSA/EC signature algorithms are set in ["
+                        + allowedSignatureAlgorithmsConfigKey
+                        + "], but no RSA/EC JWKSet is set in ["
+                        + jwkSetConfigKeyPkc
+                        + "]"
+                );
+            }
+        }
+
+        // At this point, one or both JWKSets are configured. Load and validate their JWK contents against each set of signature algorithms.
+        JWKSet jwkSetHmac = null;
+        if (isConfiguredJwkSetHmac) {
+            jwkSetHmac = JwtUtil.loadJwkSetFromString(jwkSetConfigKeyHmac, jwkSetContentsHmac);
+            JwtUtil.validateJwkSetContents(jwkSetConfigKeyHmac, jwkSetHmac, allowedSignatureAlgorithmsHmac);
+        }
+        JWKSet jwkSetPkc = null;
+        if (isConfiguredJwkSetPkc) {
+            jwkSetPkc = JwtUtil.loadJwkSetFromString(jwkSetConfigKeyPkc, jwkSetContentsPkc);
+            JwtUtil.validateJwkSetContents(jwkSetConfigKeyPkc, jwkSetPkc, allowedSignatureAlgorithmsPkc);
+        }
+        return new Tuple<>(jwkSetHmac, jwkSetPkc);
+    }
+
+    public static boolean isHttpsScheme(final URI uri) {
+        return (uri != null) && "https".equalsIgnoreCase(uri.getScheme()); // equalsIgnoreCase handles null
+    }
+
+    public static URI parseHttpsUriNoException(final String uriString) {
+        try {
+            final URI uriObj = new URI(uriString);
+            if (JwtUtil.isHttpsScheme(uriObj)) {
+                return uriObj;
+            }
+            LOGGER.trace("Invalid scheme for URI [" + uriString + "]. Only HTTPS is supported.");
+        } catch (Exception e) {
+            LOGGER.trace("Invalid URI [" + uriString + "]", e);
+        }
+        return null;
+    }
+
+    public static byte[] readUriContents(
+        final String jwkSetConfigKeyPkc,
+        final URI jwkSetPathPkcUri,
+        final CloseableHttpAsyncClient httpClient
+    ) throws SettingsException {
+        try {
+            if (JwtUtil.isHttpsScheme(jwkSetPathPkcUri)) {
+                return JwtUtil.readBytes(httpClient, jwkSetPathPkcUri, Integer.MAX_VALUE);
+            }
+            throw new SettingsException("Invalid scheme for URI [" + jwkSetPathPkcUri + "]. Only HTTPS is supported.");
+        } catch (SettingsException e) {
```

Comment:
Yes, on line 369.

Line 367 throws IOException or others related to URL retrieval.
Line 369 throws SettingsException.
Line 370 catches SettingsException and rethrows it, to avoid falling through to line 372, which would wrap SettingsException in a SettingsException.

Reply:
gotcha thanks for the explanation. Reading through this, do we need to reconsider where we are returning SettingsException in this class ? In this case, we would throw a SettingsException if for which ever reason we failed to read the JWKS ( even if the URL is correct). Just a thought for consideration

Reply:
I decided to remove the throw, catch, rethrow of that one SettingsException to simplify the code. JwtUtil.readUriContents() only has a single catch now, and it wraps the exception in a Settings Exception, so I think removing that code makes this issue go away.